### PR TITLE
Fixed #21961 -- added support for database level on delete cascade

### PR DESCRIPTION
**Pull request by tvallois**
Ticket https://code.djangoproject.com/ticket/21961

Following PR: https://github.com/django/django/pull/8661

### CODE REVIEW

**Comment by anatolzak** [comment]
Has there been any progress with this pull request?

**Comment by jacobtylerwalls** [comment]
> Has there been any progress with this pull request?

We're waiting for the new author (or anyone) to propose solutions to the final questions in [Nick Stefan's summary](https://github.com/django/django/pull/8661#issuecomment-624244307).

I'm engaging with this history for the first time, so pardon me if I misrepresent anything, but from that comment, it looks like the main remaining issues are: 
- what to do about the failing tests for invalid models.
- whether "we need to reconsider having `db_cascade` separate." Nick S. replied that no unnecessary migrations would be made in Nick P.'s scenario from the previous code review, but someone should check to see if there is test coverage to support that argument.

There are also cosmetic things the new patch author could do to help keep the patch looking great and easy to review. The version added annotations need to be bumped to 4.0, there's a merge conflict, and the commits could be squashed and the resulting one renamed following the [patch review checklist](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#patch-review-checklist). Adding Nick Stefan [as a co-author](https://docs.djangoproject.com/en/dev/internals/contributing/committing-code/#committing-guidelines) would also be a great help.

More substantively, we will also need the new author to rewrite or remove the blurb in docs/ref/models/fields.txt that refers to this ticket number and mentions this as a missing feature. Thanks.

**Comment by tvallois** [comment]
Hello,

This is my first contribution to Django so my quotes below may be inaccurate...

> what to do about the failing tests for invalid models.

I was thinking about the two failing tests on invalid models because the foreign key is pointing on an abstract and a missing model. I don't think we have to raise an error for this because this should have no repercussion during the execution of the model (because any DB_* is just an application_do_nothing + SQL statement). Nevertheless, it allows you to create constraints between a model and another table managed only by SQL.

> whether "we need to reconsider having db_cascade separate." Nick S. replied that no unnecessary migrations would be made in Nick P.'s scenario from the previous code review, but someone should check to see if there is test coverage to support that argument.

I'll have a look for any test coverage for this. If none exists, i'll try to create one. I have the same opinion as Nick S. to keep one single kwarg to manage the cascade deletion.

File: django/db/models/fields/related.py [hunk 1]
```diff
@@ -879,6 +882,57 @@ def _check_on_delete(self):
...
+                )
+            ]
+        elif on_delete in [CASCADE, RESTRICT, SET_DEFAULT, SET_NULL] and (
+            any(  # field_A <- CASCADE <- field_B <- DB_CASCADE <- field_C
+                getattr(field_B.remote_field, 'on_delete', None) and
+                getattr(field_B.remote_field, 'on_delete') in [
+                    DB_CASCADE, DB_RESTRICT, DB_SET_NULL
+                ]
+                for field_B in self.remote_field.model._meta.fields
+            )
```
Comment (jacobtylerwalls) [inline django/db/models/fields/related.py:924]: First step is to just loosen this condition to not choke when `_meta` doesn't exist.

File: docs/releases/4.0.txt [hunk 2]
```diff
@@ -311,6 +311,11 @@ Models
   allows specifying a value to be returned when the queryset (or grouping)
   contains no entries, rather than ``None``.
 
+* The ``on_delete`` argument for ``ForeignKey`` and ``OneToOneField`` now accepts
+  ``DB_CASCADE``, ``DB_SET_NULL``, and ``DB_RESTRICT``. These
+  will behave like ``DO_NOTHING`` in Django, but leverage the ``ON DELETE`` SQL
+  constraint equivalents.
+
```
Comment (jacobtylerwalls) [inline docs/releases/4.0.txt:318]: The 4.0 feature freeze passed earlier this month, so we should move to 4.1.txt

File: docs/ref/models/fields.txt [hunk 3]
```diff
@@ -1638,6 +1638,31 @@ The possible values for :attr:`~ForeignKey.on_delete` are found in
     integrity, this will cause an :exc:`~django.db.IntegrityError` unless
     you manually add an SQL ``ON DELETE`` constraint to the database field.
 
+* .. attribute:: DB_CASCADE
+
+    .. versionadded:: 4.0
+
+    Django takes no direct action, behaves exactly like ``DO_NOTHING``,
+    but also sets ``ON DELETE CASCADE`` as a SQL constraint. The cascade
```
Comment (jacobtylerwalls) [inline docs/ref/models/fields.txt:1646]: bump to 4.1,  and "an" SQL (ess-queue-ell in Django docs convention). same directly below.

**Review by jacobtylerwalls — COMMENTED** [review]
Thanks for the updates and for the summary. The failing tests, once we get them to pass, will show that we already raise system checks for those invalid models, so I'm not currently aware of a reason we would need additional checks.

When you've got the tests passing, I suggest unchecking _Patch needs improvement_ on the ticket to get more eyes on it and finalize the TODOs. Thanks!

**Comment by felixxm** [comment]
@tvallois Do you have time to keep working on this?

**Comment by felixxm** [comment]
Closing due to inactivity.

**Pull request closed without merging**